SparseGrids/tsgCoreOneDimensional.cpp

```diff
@@ -392,6 +393,168 @@ const char* OneDimensionalMeta::getHumanString(TypeOneDRule rule){
+        const std::vector<double> &ref_points) {
+
+    // Initialize and compute the integral weights corresponding to weight_fn.
+    std::vector<std::vector<double>> root_cache(n);
+    std::vector<double> roots;
+    #ifdef Tasmanian_ENABLE_BLAS
+    assert(ref_points.size() == ref_weights.size());
+    std::vector<double> integral_weights(ref_points.size());
+    for (size_t j=0; j<ref_points.size(); j++) {
+        integral_weights[j] = weight_fn(ref_points[j]) * ref_weights[j];
+    }
+
+    // Compute the roots incrementally.
+    std::vector<double>
+            poly_m1_vals(ref_points.size(), 0.0),
+            poly_vals(ref_points.size(), 1.0);
+    std::vector<double> alpha, beta;
+    for (int i=0; i<n; i++) {
+        // Form the tridiagonal vectors, alpha and beta, of the Jacobi matrix.
+        double alpha_numr = 0.0;
+        double alpha_denm = 0.0;
+        double beta_numr = 0.0;
+        double beta_denm = 0.0;
+        for (int j=0; j<ref_points.size(); j++) {
+            alpha_numr += ref_points[j] * (poly_vals[j] * poly_vals[j]) * integral_weights[j];
+            alpha_denm += (poly_vals[j] * poly_vals[j]) * integral_weights[j];
+            beta_numr += (poly_vals[j] * poly_vals[j]) * integral_weights[j];
+            beta_denm +=  (poly_m1_vals[j] * poly_m1_vals[j])  * integral_weights[j];
+        }
+        alpha.push_back(alpha_numr / alpha_denm);
+        if (i >= 1) {
+            beta.push_back(beta_numr / beta_denm);
+        }
+        // Compute the roots. alpha and beta need to be copied because the
+        // tridiagonal eigensolver generates the eigenvalues in place and
+        // destroys some of its inputs.
+        roots = alpha;
+        std::vector<double> offdiag(n-1);
+        for (int k=0; k<i; k++) {
+            offdiag[k] = sqrt(beta[k]);
+        }
+        TasmanianTridiagonalSolver::getSymmetricEigenvalues(i+1, roots, offdiag);
+        if (roots.size() % 2 == 1) {
+            roots[(roots.size() - 1) / 2] = 0.0; // Zero out the center for stability.
+        }
+        root_cache[i] = roots;
+        // Update the values of the polynomials at ref_points.
+        poly_m1_vals = poly_vals;
+        std::transform(ref_points.begin(), ref_points.end(), poly_vals.begin(),
+                       [&roots, i](double x){return poly_eval(roots, x);});
+    }
+    #endif
+    return root_cache;
+}
+
+// Get Exotic Gauss-Legendre points and weights. For the case where the
+// integrand F(x) is of the form:
+//     F(x) := f(x) * [weight_fn(x) - shift] + shift * f(x)
+double lagrange_eval(int idx, const std::vector<double> roots, double x) {
+    double eval_value = 1.0;
+    for (size_t i=0; i<roots.size(); i++) {
+        if (i != idx) {
+            eval_value *= (x - roots[i]) / (roots[idx] - roots[i]);
+        }
+    }
+    return eval_value;
+}
+void OneDimensionalExoticQuad::getExoticGaussLegendreCache(
+        std::vector<std::vector<double>> &weights_cache,
```

Comment:
All outputs should follow all inputs, inputs come first.

Reply:
Changed.